### PR TITLE
feat: Support overwrites for interface bindings and endpoints

### PR DESCRIPTION
**Pull request by timo0**
Fixes https://github.com/hashgraph/hedera-services/issues/17075

Added configuration capabilities to customize network settings for nodes running in containerized environments:

* Configure interface bindings for receiving local Gossip traffic
* Override peer endpoint addresses for outgoing Gossip connections

**Implementation Decisions**

* Configuration is managed through a new `GossipConfig` platform component
* Network settings are stored in a dedicated `overwrites.yaml` file
* Single nodes cannot bind to multiple interfaces
* Multiple node bindings can be specified in the same configuration file

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/InterfaceBinding.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.gossip.config;
+
+import java.net.InetAddress;
+
+public record InterfaceBinding(Long nodeId, InetAddress hostname, int port) {}
```
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/InterfaceBinding.java:21]: null checks?

File: platform-sdk/sdk/overwrites.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+gossip:
+  interfaceBindings:
```
[the remark this answers is not in the document]
Comment (timo0) [reply platform-sdk/sdk/overwrites.yaml]: sorry was committed by accident - deleted

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/GossipConfig.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+     * Returns the interface binding for the given node ID.
+     * <p>
+     *     <b>Note:</b> If there are multiple interface bindings for the same node ID, only the first one will be
+     *     returned.
+     *</p>
+     *
+     * @param nodeId the node ID
+     * @return optional of the interface binding, empty if not found
+     */
+    public Optional<InterfaceBinding> getInterfaceBinding(long nodeId) {
```
[the remark this answers is not in the document]
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/GossipConfig.java]: renamed, also the variable

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/InterfaceBinding.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.gossip.config;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.net.InetAddress;
+import java.util.Objects;
+
+public record InterfaceBinding(@NonNull Long nodeId, @NonNull InetAddress hostname, int port) {
```
[the remark this answers is not in the document]
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/InterfaceBinding.java]: changed, thanks :)

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/InterfaceBinding.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    public InterfaceBinding {
+        Objects.requireNonNull(nodeId, "nodeId must not be null");
+        Objects.requireNonNull(hostname, "hostname must not be null");
```
[the remark this answers is not in the document]
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/InterfaceBinding.java]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/SocketFactory.java [hunk 6]
```diff
@@ -48,27 +52,44 @@ static boolean isIpTopInRange(final int ipTos) {
...
             serverSocket.setOption(java.net.StandardSocketOptions.IP_TOS, socketConfig.ipTos());
         }
-        final InetSocketAddress endpoint = new InetSocketAddress(InetAddress.getByAddress(ALL_INTERFACES), port);
+        final InetSocketAddress endpoint = new InetSocketAddress(interfaceBinding.hostname(), interfaceBinding.port());
         serverSocket.setReuseAddress(true);
         serverSocket.bind(endpoint); // try to grab a port on this computer
         // do NOT do clientSocket.setSendBufferSize or clientSocket.setReceiveBufferSize
         // because it causes a major bug in certain situations
-
-        serverSocket.setSoTimeout(socketConfig.timeoutServerAcceptConnect());
```
[the remark this answers is not in the document]
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/SocketFactory.java]: accident, added back - thanks

**Review by edward-swirldslabs — APPROVED** [review]

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM! Thanks @timo0

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/InterfaceBindingConverter.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/config/InterfaceBindingConverter.java:6]: Does this class name need to be changed?

**Review by artemananiev — APPROVED** [review]
Looks good from platform-data perspective

File: platform-sdk/swirlds-config-extensions/src/test/resources/test.yaml [hunk 8]
```diff
@@ -1,5 +1,5 @@
 gossip:
-  interfaceBindings:
+  networkEndpoints:
     - { "nodeId": 0, "hostname": "10.10.10.1", "port": 1234 }
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-config-extensions/src/test/resources/test.yaml:3]: This format seems too hacky for my taste. As it is already in the repo, this comment is not a blocker,
but I'm curious why we didn't use a complete yaml list like:
```
networkEndpoints:
  - nodeId: 0
    hostname: 10.10.10.1
    port: 1234
  - nodeId: 0
    hostname: 10.10.10.2
    port: 1234
```

**Review by mxtartaglia-sl — APPROVED** [review]

**Pull request merged**